### PR TITLE
[RSI] Bounded provider wait-for-usage, transient-unavailability pings, and optional user-defined backup model

### PR DESCRIPTION
**Pull request by sethkarten**
> **User-directed feature** (2026-09-13):
> "add a feature that waits for usage to come back (eg codex sub) or ping with exponential backoff if the provider is unavailable or maybe the ability to set a backup model in these cases (user defined, default no backup)"

This PR was authored autonomously as part of a recursive self-improvement program (RSI). Source: cycle B wave item 11, `/Users/milkkarten/rsi/reports-cycle-b/builder-specs/provider-resilience.txt` (+ `SYNTHESIS.md`).

## Why (evidence from this program's own operation)

- **2026-09-13T04:17:23Z**: a 2-second 404 burst on `internal/glm-5.3-fast` killed every child session whose stream was active at that instant (7 `provider stream failure` entries in the daemon log). Root amplifier: `classifyStreamFailure` maps 404 to `invalid_request`, which the retry loop treats as **permanent** — no retry, no fallback; sessions died mid-turn. (The same burst also produced 503s, which retried 3× with 2/4/8s delays and then died.)
- **Quota gating is a standing blocker** (program STATE.md: ChatGPT sub limit ~81h, anthropic OAuth expired): agents currently have no way to wait productively for a quota reset; the codex 429 is `You have hit your ChatGPT usage limit ... Try again in ~7272 min.` — today it either burns 3 quick retries or fails via the `maxRetryDelayMs` cap with no waiting.
- Harness lessons applied: cost questions are stop signals, abort criteria are mandatory — **all waits here are bounded** (max attempts + max total duration), never infinite.

## What

1. **Error classifier for recovery routing** (`providerWaitClass` in `provider-retry.ts`): quota/subscription exhaustion (429, `rate_limit`) vs transient unavailability (5xx, overload, network/unknown, **and 404s**) vs permanent (auth, permission, refusal, safety, malformed, non-404 invalid requests). Reuses the existing `StreamFailureKind` typing; a 404 is no longer treated as a permanent rejection anywhere (`isPermanentProviderFailureKind` now takes the status).
2. **Wait-for-usage mode**: on quota failures the session enters a bounded wait — ping with exponential backoff + jitter (1s doubling to a 5m ceiling), resuming automatically when the provider recovers. If the provider reports a reset time (Retry-After header or "Try again in ~7272 min" text), the resume is scheduled exactly then (no pointlessly early pings); if that reset is beyond the configured wait bound, the wait gives up immediately with an informative error. Surfaced in the status line with attempts + a next-check countdown (same loader the existing retry UI already renders, extended with wait/backup wording). Abort bounds: `maxAttempts` (30) and `maxWaitMs` (15m), both configurable; `retry.enabled: false` disables the whole machinery.
3. **Transient-unavailable mode**: after the existing quick retries (3×, 2s/4s/8s) are exhausted on 5xx/overload/network/404, the same bounded ping loop takes over instead of killing the session.
4. **Backup model** (`providerBackupModel`, **default: none** — never a surprise cost): a user-defined `"provider/model-id"` reference the failed turn is routed to while the primary is quota-blocked or unavailable. Explicit "retrying on backup model X" indicator, `model_change` entries in the session log for every primary->backup and backup->primary transition, and automatic return to the primary (the next turn probes the primary; a still-blocked primary routes straight back to the backup). Unresolvable/unauthenticated backup references fall back to the bounded wait — no crash, no silent behavior.
5. **Telemetry**: transitions are session-logged (`model_change` entries) and the retry events now carry `reason` (`usage` | `unavailable` | `backup`), `backupModel`, and `restoredModel`, so incidents are debuggable after the fact.

## Change

- `packages/coding-agent/src/core/provider-retry.ts`: `providerWaitClass`, `ProviderWaitPolicy` + defaults, `providerWaitPingDelay` (pure exponential + cap), `providerWaitJitter` (±25%, injectable rng), `providerWaitDecision` (bounded decision incl. reset-aware resume and both abort bounds), `parseProviderResetMs` (real codex-plan shape), `providerStreamFailureStatus`; 404 now non-permanent in `isPermanentProviderFailureKind`.
- `packages/coding-agent/src/core/settings-manager.ts`: `retry.provider.waitForUsage` settings (enabled/baseDelayMs/maxDelayMs/maxAttempts/maxWaitMs), `providerBackupModel`, normalized getters.
- `packages/coding-agent/src/core/agent-session.ts`: retry-loop integration (backup routing, quota wait, transient wait after quick-retry exhaustion), the shared park/sleep/continue tail extracted into `_retryAfterDelay` (used by all three paths), backup switch/restore with thinking-level and service-tier clamping.
- `packages/coding-agent/src/modes/agent-connection/types.ts` + `interactive-mode.ts`: optional retry-event fields + status-line wording ("Waiting for provider usage to recover (2/30), next check in 8s...", "retrying on backup model X", "Primary provider recovered — back on Y").
- Docs: `docs/settings.md` (new "Wait-for-usage and provider recovery" section); changelog fragment `.changes/provider-wait-backup-model.md`.

Deliberate behavior changes for the quota class (defaults): quick retries (which cannot cover usage windows) are replaced by the bounded wait; the `maxRetryDelayMs` server-request cap no longer governs quota failures when the wait loop is enabled — its own `maxWaitMs` bound governs (set `retry.provider.waitForUsage.enabled: false` to restore the exact previous behavior; two existing tests updated accordingly, keeping the cap's coverage under the disabled configuration).

## Validation

- `npm run check` (biome + tsgo + installer + browser-smoke): clean.
- Targeted vitest: `provider-retry.test.ts` (19), `settings-manager.test.ts` (41), `suite/agent-session-retry-events.test.ts` (43 incl. 11 new), `suite/regressions/4491-provider-stale-after-401.test.ts` (13), `suite/regressions/3317-network-connection-lost-retry.test.ts` (1), `packages/ai` `stream-failure.test.ts` + `provider-retry-ownership.test.ts` (44) — all green.
- New tests are model-free and deterministic: classifier cases from the real error shapes above (codex 429 text, prime-inference 503/404, Anthropic 529), backoff schedule with injectable rng (exponential growth, jitter bounds, cap, both abort bounds, reset parsing incl. the exact codex string), backup routing on/off, unresolvable-backup fallback, return-to-primary probing on the next turn.
- Pre-existing failures on origin/main (agent-session-queue auto-refine set, agent-session-runtime rlmDepth) are untouched by this change; verified by stash/run/restore on a clean tree.

## Notes for reviewers

- **Overlap with #795** ("fail closed on provider quota exhaustion"): #795 makes quota exhaustion terminal (circuit breaker, stop until an explicit user prompt). This PR implements the user's opposite-leaning request: bounded waits and an opt-in backup. The two can compose (e.g. fail-closed by default, wait/backup only when explicitly configured), but this PR deliberately does not pre-decide that; it builds on current main, does not touch `classifyStreamFailure` or the `StreamFailureKind` union (rebase should be small either way), and `providerWaitClass` is forward-compatible with a future dedicated `quota` kind.
- This touches `agent-session.ts` (unavoidable: the retry loop lives there); the edit is one cohesive region (`_handleRetryableError` + extracted tail + small state fields) to limit rebase pain against the open modularization stack.
- Scope kept to the session retry loop; one-shot completion consumers (`completeWithProviderRetry`: compaction, refinement, side questions) keep the existing quick policy — only the 404-classification fix applies there.

---

> [!NOTE]
> **Medium Risk**
> Changes default retry behavior for quota failures and can switch models when `providerBackupModel` is set; core session retry logic is large but bounded and heavily tested.
> 
> **Overview**
> Adds **provider recovery** behavior so sessions are less likely to die mid-turn on quota limits or transient outages.
> 
> **Wait-for-usage** (default on under `retry.provider.waitForUsage`): quota/`rate_limit` failures enter a **bounded** ping loop (exponential backoff with jitter, configurable max attempts and total wait) instead of stopping after quick retries or the `maxRetryDelayMs` cap. When Retry-After or parsed reset text fits the budget, the next attempt is scheduled at that time; otherwise the wait aborts with a clear error.
> 
> After quick retries exhaust on **transient** failures (5xx, overload, unknown, and **404 routing blips**), the same wait loop runs instead of failing immediately. Structured **404** `invalid_request` is no longer treated as permanently non-retryable in `isPermanentProviderFailureKind` / SDK retry helpers.
> 
> Optional **`providerBackupModel`**: when set and resolvable, quota/transient errors can **immediately retry on the backup** (status line + session `model_change` logging), then **restore** the primary model, thinking level, and service tier on success, cancel, or failure. Unresolvable backups fall back to the wait loop.
> 
> Retry events and the interactive UI gain `reason` (`usage` | `unavailable` | `backup`), backup/restored model fields, and wait countdown copy. Settings docs and tests cover defaults, abort bounds, backup routing, and stale-continue cancel edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ee1f77e98ca2cc7de525806a591897ff5192587f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add bounded provider recovery waits, backup-model routing, and retryable 404 handling in `AgentSession`
> - Quota and transient provider failures now enter a bounded wait loop with exponential backoff, jitter, and provider-reported reset delays, configurable via `retry.provider` settings (enabled by default).
> - Non-permanent provider failures can route to a user-configured `providerBackupModel`; the session saves and restores the primary model, thinking level, and service tier on success or terminal failure.
> - Structured 404 invalid-request failures are now classified as retryable transient failures instead of permanent errors in `isPermanentProviderFailureKind` and `completeWithProviderRetry` in [provider-retry.ts](https://github.com/PrimeIntellect-ai/prime-agent/pull/2280/files#diff-6b31923dc9faefc23eb791f813ba9596050d1fd8c0c238f2e6f1738de59691fc).
> - Retry events (`auto_retry_start`/`auto_retry_end`) carry backup, usage, and unavailable reasons plus restored-model metadata; interactive mode surfaces these in the status line.
> - Risk: 404 invalid-request failures previously terminal now retry; existing tests disable wait-for-usage to preserve legacy behavior. Default-enabled recovery waits may delay terminal failures by up to the configured duration bounds.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized ee1f77e.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/core/settings-manager.ts [hunk 1]
```diff
@@ -957,6 +974,24 @@ export class SettingsManager {
...
+			baseDelayMs: bound(wait?.baseDelayMs, 1000),
+			maxDelayMs: bound(wait?.maxDelayMs, 300_000),
+			maxAttempts: bound(wait?.maxAttempts, 30),
+			maxWaitMs: bound(wait?.maxWaitMs, 900_000),
+		};
+	}
+
+	getProviderBackupModel(): string | undefined {
+		const reference = this.settings.providerBackupModel;
+		return reference?.trim() ? reference.trim() : undefined;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/settings-manager.ts]: Valid — fixed in bbcf11c91. `getProviderBackupModel()` now checks `typeof reference === "string"` and treats any non-string JSON value as unset, so a corrupted setting can no longer throw into the retry path. Covered by `settings-manager.test.ts` ("treats a non-string providerBackupModel value as unset instead of throwing").

File: packages/coding-agent/src/core/agent-session.ts [hunk 2]
```diff
@@ -1296,6 +1308,11 @@ export class AgentSession {
 	private _retryPromise: Promise<void> | undefined = undefined;
 	private _retryResolve: (() => void) | undefined = undefined;
 	private _retryAuthFailureSources: AuthSourceToken[] = [];
+	/** Ongoing wait-for-recovery state: pings issued and when the wait started. */
+	private _providerWait: { attempts: number; startedAtMs: number } | undefined = undefined;
+	/** Set while turns are routed to the user-configured backup model. */
+	private _backupModel: { backup: Model<any>; primary: Model<any>; thinkingLevel: ThinkingLevel } | undefined =
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Valid — fixed in bbcf11c91. `_backupModel` now stores the prior effective `serviceTier` captured at switch time, and `_restorePrimaryModelAfterBackup` reclamps with that saved tier instead of the backup-clamped current state, so a `priority` primary is no longer downgraded to `default`. Covered by the service-tier regression test in `agent-session-retry-events.test.ts` ("restores the saved service tier after a backup retry, not the backup-clamped one").

File: packages/coding-agent/src/core/agent-session.ts
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:1]: Valid — fixed in bbcf11c91, including the same-class gaps beyond the two paths flagged here. All terminal retry paths that zero `_retryAttempt` now call `_restorePrimaryModelAfterBackup()` first: the sleep-abort catch in `_retryAfterDelay`, the no-controller branch of `abortRetry()`, the max-retry give-up, the server-request-cap give-up, and the wait-bound abort. The restore also clears `_backupModel` unconditionally, so no later turn (including successes outside the retry loop) can run silently on the backup. Regression tests: "restores the primary model when a backup-model retry is cancelled mid-wait", "...when quick retries exhaust on the backup model", "...when the bounded wait aborts on the backup model".

File: packages/coding-agent/src/core/agent-session.ts [hunk 3]
```diff
@@ -11748,6 +11823,130 @@ export class AgentSession {
...
+				type: "auto_retry_end",
+				success: false,
+				attempt: pingAttempt - 1,
+				finalError: `${decision.message}: ${message.errorMessage || "unknown error"}`,
+			});
+			this._retryAttempt = 0;
+			this._providerWait = undefined;
+			this._retryAuthFailureSources = [];
+			this._resolveRetry();
+			return false;
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:11941]: Valid — fixed in bbcf11c91. Every terminal retry path that zeroes `_retryAttempt` now calls `_restorePrimaryModelAfterBackup()` before returning (sleep-abort catch, no-controller `abortRetry()`, max-retry give-up, server-request-cap give-up, wait-bound abort), and the helper clears `_backupModel` unconditionally, so no later turn — success or failure, in or out of the retry loop — can stay silently on the backup. Deterministic regression tests in `agent-session-retry-events.test.ts` cover the cancel, give-up, and wait-abort cases.

File: packages/coding-agent/src/core/agent-session.ts [hunk 4]
```diff
@@ -11748,6 +11823,130 @@ export class AgentSession {
...
+	 * switched models meanwhile. Returns the restored "provider/model-id"
+	 * reference for the retry-end event.
+	 */
+	private _restorePrimaryModelAfterBackup(): string | undefined {
+		const backup = this._backupModel;
+		this._backupModel = undefined;
+		if (!backup || !modelsAreEqual(this.model, backup.backup)) return undefined;
+		this.agent.state.model = backup.primary;
+		this.agent.state.thinkingLevel = backup.thinkingLevel;
+		this._clampServiceTierForModel();
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts]: Valid — fixed in bbcf11c91 (same root cause as the Macroscope service-tier finding). `_backupModel` now stores the prior effective `serviceTier`, and `_restorePrimaryModelAfterBackup` reclamps with the saved tier rather than re-deriving from the backup-clamped current value, so a `priority` primary returns to `priority`. Regression test: "restores the saved service tier after a backup retry, not the backup-clamped one".

File: packages/coding-agent/src/core/agent-session.ts [hunk 5]
```diff
@@ -11713,11 +11796,14 @@ export class AgentSession {
...
 			this._retryAttempt = 0;
 			this._retryAbortController = undefined;
+			this._providerWait = undefined;
+			const restoredModel = this._restorePrimaryModelAfterBackup();
 			this._emit({
 				type: "auto_retry_end",
 				success: false,
 				attempt,
 				finalError: "Retry cancelled",
+				...(restoredModel ? { restoredModel } : {}),
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:11806]: Valid — fixed in 2531b2b73. The `agent.continue()` rejection handler in `_retryAfterDelay` now calls `_restorePrimaryModelAfterBackup()` alongside the other terminal exits, so a backup-routed turn whose continue never starts restores the primary and reports `restoredModel` in the `auto_retry_end` event. Regression test: "restores the primary model when the scheduled backup retry continue cannot run".

File: packages/coding-agent/src/core/agent-session.ts [hunk 6]
```diff
@@ -11756,13 +11977,16 @@ export class AgentSession {
 		if (this._retryAttempt > 0) {
 			this._autoCompactionAbortController?.abort();
 			this._cancelPostCompactionContinue();
+			const restoredModel = this._restorePrimaryModelAfterBackup();
 			this._emit({
 				type: "auto_retry_end",
 				success: false,
 				attempt: this._retryAttempt,
 				finalError: "Retry cancelled",
+				...(restoredModel ? { restoredModel } : {}),
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply packages/coding-agent/src/core/agent-session.ts:11991]: Valid — fixed in 2531b2b73. The scheduled retry continue is now guarded before starting: `if (this._retryGeneration !== retryGeneration || !this.isRetrying) return;` precedes `agent.continue()`. Since `abortRetry()` resolves the retry (bumping `_retryGeneration`), a cancel landing between the wait and the scheduled start no longer re-issues the turn at all — previously the queued continue could fire onto the restored, quota-blocked primary and start a fresh backup cycle the user just cancelled. Regression test: "does not re-issue a wait retry cancelled between the delay and the scheduled continue", which aborts deterministically in that window (microtasks run before the scheduled timer).

**Review by xeophon — APPROVED** [review]
